### ISSUE TITLE
REST API Authentication does not work if wp_get_current_user() is called before WooCommerce is loaded

### ISSUE THREAD

**Issue by om4james** [woocommerce/woocommerce#26847]
**Prerequisites (mark completed items with an [x]):**
- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
WooCommerce's REST API authentication relies on the `determine_current_user` function in order to perform its authentication ([ref](https://github.com/woocommerce/woocommerce/blob/4.2.1/includes/class-wc-rest-authentication.php#L41)). This filter runs whenever WordPress' `wp_get_current_user()` function is executed for the very first time during a request.

If a site has a plugin or code that calls `wp_get_current_user()` early during bootstrap (rather than waiting until `plugins_loaded`, then REST API requests return an HTTP 401 response:

```
{"code":"woocommerce_rest_cannot_view","message":"Sorry, you cannot list resources.","data":{"status":401}}
```

**Expected behavior**
I expect REST API requests to authenticate correctly.

**Actual behavior**
HTTP 401 reponses are received, even though a valid key and secret are provided via the URL parameters.

Our WooCommerce Zapier extension uses the WooCommerce REST API. We have approximately 2-3% of our user base unable to authenticate because their REST API returns HTTP 401 responses even though valid keys and secrets are provided.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. Create a REST API consumer key and secret via wp-admin --> WooCommerce --> Settings --> Advanced -->  REST API.
2. Load `https://example.com/wp-json/wc/v3/orders/?consumer_key=ck_***&consumer_secret=cs_***` in your browser or API client such as Insomnia.  Ensure the `consumer_key=` and `consumer_secret=` parameters in the URL are updated to use your newly created key and secret from step 1. Confirm that  HTTP 200 OK response is received, which demonstrates that HTTP basic authentication via query parameters is working as expected.
3. Create a `wp-content/mu-plugins/file.php` file with the following in it: 
```
require ABSPATH . WPINC . '/pluggable.php';
wp_get_current_user();
```
Alternatively, you could add the above code to any active plugin's main PHP file, provided that the plugin folder is alphabetically before `woocommerce`.
The above code causes `wp_get_current_user()` to be executed before WooCommerce is bootstrapped.
4. Refresh your browser/client from step 2. A HTTP 401 `woocommerce_rest_cannot_view` `Sorry, you cannot list resources` response is received.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
N/A
</details>

**Comment by om4james** [comment]
To give a little more context:

WooCommerce Zapier version 2 uses the WooCommerce REST API exclusively.  

We already have thousands of users using it successfully, but we've also seen a surprising amount of users not able to authenticate with the WooCommerce REST API even though a valid consumer key and secret are specified.

Instead, they receive an HTTP 401 response with a `Sorry, you cannot list resources` error message.

We originally tried using [HTTP Basic Auth for consumer keys and secrets](https://woocommerce.github.io/woocommerce-rest-api-docs/#authentication-over-https), and it turns out HTTP basic authentication isn't supported by a surprising number of web hosts out there.

So we changed to using query string parameters for providing consumer keys and secrets, and a lot more users are now successfully authenticating.

However, we still see a small percentage receiving unexplained HTTP 401 responses and this bug report I hope outlines a possible avenue where auth can fail.

My hope is that we can find a way for WooCommerce core to more robustly authenticate REST API requests, even if a site is using one of many plugins that may do something strange.

Would anyone be willing to discuss this in more detail and put our heads together to come up with a more robust solution?

I think it should be possible for WooCommerce to continue attempting to authenticate during `determine_current_user`, but also have a fallback so that if the auth mechanism doesn't load during that filter, then it loads later in the REST API request lifecyle.

Thank you!

**Comment by ObliviousHarmony** [comment]
Hello @om4james,

Thank you very much for such a detailed report. I can confirm that I am able to replicate the bug as described. When I think about possible solutions to this, the first thing that comes to mind is setting the `$current_user` global to `null` if it has been set to `new WP_User( 0 )` by an earlier call. I wonder what ramifications this may have though. After looking through the REST API, this seems like it might be the best course of action.

What are your thoughts?

**Comment by om4james** [comment]
Hi @ObliviousHarmony, thanks for getting back to me!

I [experimented with a possible solution to this here](https://github.com/woocommerce/woocommerce/compare/master...om4james:fix/26847-rest-api-auth).

In the `WC_REST_Authentication::check_user_permissions()` function, if the user isn't set and the `WC_REST_Authentication::authenticate()` function hasn't yet been executed, then it calls it.

Do you think that kind of approach may be a good way to solve this?

Unfortunately, the `WC_REST_Authentication` class has 1% code coverage, so none of this is tested as far as I'm aware.

Thank you :)

**Comment by om4james** [comment]
Hi @ObliviousHarmony, 

Do you think we're on the right track with a solution to this?

Any thoughts and/or guidance would be appreciated, as this issue affects a small portion of users using the WooCommerce REST API.

Thanks,

**Comment by ObliviousHarmony** [comment]
Hi @om4james,

Your solution is alright, though, I'll note you don't need to call `is_request_to_rest_api()` because `authenticate()` already does so. I wonder if given the restrictions we've placed on authentication if we should stop listening to `determine_current_user` and authenticate on `rest_api_init` instead? There's an open [Trac ticket for Core](https://core.trac.wordpress.org/ticket/46586) about the limitations of `determine_current_user` as well.

**Comment by om4james** [comment]
Hi @ObliviousHarmony , I like the idea of re-thinking when the authentication is performed. Moving it from `determine_current_user` to `rest_api_init` instead makes sense.

However, would it cause a backwards compatibility problem for anyone that is relying on the current behaviour? I imagine it could if someone has custom code that is relying on it occurring during `determine_current_user`.

This was my main reason for adding the "fallback" to the `WC_REST_Authentication::check_user_permissions()` function, so current behaviour doesn't change at all.

**Comment by om4james** [comment]
Hi @ObliviousHarmony,

It would be great to prioritise this issue, so it can be included in the next WooCommerce release.

We've seen 4 WooCommerce Zapier users in August report this issue to us, and there is no doubt many others who haven't contacted us about it.

It's likely affecting a small percentage of others out there who are using the WooCommece REST API and are unable to authenticate due to this.

I'm happy to work on a fix for this ASAP, but I'd like some guidance first so we both agree on the best solution first.

Thank you.

James

**Comment by ObliviousHarmony** [comment]
Sorry for the late reply @om4james,

Based on the [decsription of the rest_authentication_errors filter](https://developer.wordpress.org/reference/hooks/rest_authentication_errors/) it does seem like it is fine to use it for general authentication. My thinking is that instead of extending `check_user_permissions` we should add another hook to `rest_authentication_errors` before the current one. We can then try to authenticate if `$this->error` is null, `wp_get_current_user_id()` is zero, and the input does not indicate it was already handled. This should both avoid any possible `determine_current_user` loops as well as only trigger in the error case from this issue.

**Comment by om4james** [comment]
Thank you for the feedback @ObliviousHarmony.

I have just implemented the changes and submitted PR #27587

**Comment by jacobraccuia** [comment]
I ran into a problem trying to call a WC endpoint internally, using `wp_remote_get`, inside of a `save_post` action.  Authentication was failing inside the aforementioned `authentication_fallback`.  

I tried to debug all of the variables and found that my user ID was already set during the initial `authenticate` method, so the fallback was triggering. `get_current_user_id` was returning a value and therefore it never attempted to authenticate and returned a null error.

I am trying to understand the code better, and I can't seem to get why this fallback method is ONLY authenticating if the user_id is 0.  In my case, the user_id was set, and no authentication had passed yet.

I likely have an obscure use case, but essentially I am trying to call API methods internally to regenerate the data for the WP REST CACHE API plugin.  This plugin regenerates the cache on cron jobs and failing authentication.  I tried to manually trigger it when posts were saved, and it started to fail.

To currently bypass the problem, I am sending a nonce through wp_remote_get and adding a filter to rest_authentication_errors and passing true if the nonce is verified.
